publications/radstack/radstack.tex

```diff
@@ -0,0 +1,1148 @@
+node. The node starts at 13:37 and, with a 10 minute window period, will only
+accept events for a window between 13:27 and 13:47.  When the first events are
+ingested, the node announces that it is serving a segment for an
+interval from 13:00 to 14:00. Every 10 minutes (the persist period is
+configurable), the node will flush and persist its in-memory buffer to disk.
+Near the end of the hour, the node will likely see events for 14:00 to 15:00.
+When this occurs, the node prepares to serve data for the next hour and creates
+a new in-memory buffer.  The node then announces that it is also serving a
+segment from 14:00 to 15:00.  At 13:10, which is the end of the hour plus the
+window period, the node begins the hand off process.
+
+\begin{figure*}
+\centering
+\includegraphics[width = 4.5in]{realtime_timeline}
+\caption{The node starts, ingests data, persists, and periodically hands data
+off. This process repeats indefinitely. The time periods between different
+real-time node operations are configurable.}
+\label{fig:realtime_timeline}
+\end{figure*}
+
+\subsection{Hand off}
+Real-time nodes are designed to deal with a small window of recent data and
+need periodically hand off segments they’ve built. The hand-off process first
+involves a compaction step. The compaction process finds all the segments that
+were created for a specific interval of time (for example, all the segments
+that were created by intermediate persists over the period of an hour). These
+segments are merged together to form a final immutable segment for handoff. 
+
+Handoff occurs in a few steps. First, the finalized segment is uploaded to a
+permanent backup storage, typically a distributed file system such as S3
+\cite{decandia2007dynamo} or HDFS \cite{shvachko2010hadoop}, which Druid refers
+to as “deep storage”. Next, an entry is created in the metadata store
+(typically a RDBMS such as MySQL) to indicate that a new segment has been
+created. This entry in the metadata store will eventually cause other nodes in
+the Druid cluster to download and serve the segment. The real-time node
+continues to serve the segment until it notices that the segment is available
+on Druid historical nodes, which are nodes that are dedicated to serving
+historical data. At this point, the segment is dropped and unannounced from the
+real-time node. The entire handoff process is fluid; data remains continuously
+queryable throughout the entire handoff process. Segments created by real-time
+processing are versioned by the start of the segment granularity interval.
+
+\subsection{Batch Data Ingestion}
+The core component used by real-time ingestion is an index that can be
+incrementally populated and finalized to create an immutable segment. This core
+component is shared across both real-time and batch ingestion. Druid has built
+in support for creating segments by leveraging Hadoop and running MapReduce
+jobs to partition data for segments.  Events can be “streamed” directly from
```

Comment:
This last sentence seems like a non-sequitur.  It needs more context to make sense.